bigint-asm/u256.c

```diff
@@ -0,0 +1,141 @@
+#include <stdint.h>
+
+uint8_t u256add(uint64_t *first, uint64_t *second) {
+  uint8_t overflow;
+
+  asm (
+       "addq %2, (%1)\n  "
+       "adcq %3, 8(%1)\n  "
+       "adcq %4, 16(%1)\n  "
+       "adcq %5, 24(%1)\n  "
+       "setc %0"
+       : "=r"(overflow)
+       : "r"(first),
+         "r"(second[0]),
+         "r"(second[1]),
+         "r"(second[2]),
+         "r"(second[3])
+       );
+
+  return overflow;
+}
+
+uint64_t u256mul(uint64_t *first, uint64_t *second, uint64_t *out) {
+  register uint64_t overflow asm("rcx");
+  uint64_t result[4];
+
+  // TODO: Could use movcc instead of jumpcc?
+  asm (
+    "movq %4, %%rax\n  "
+    "mulq %8\n  "
+    "movq %%rax, %0\n  "
+    "movq %%rdx, %1\n  "
+
+    "movq %4, %%rax\n  "
+    "mulq %9\n  "
+    "addq %%rax, %1\n  "
+    "adcq $0, %%rdx\n  "
+    "movq %%rdx, %2\n  "
+
+    "movq %4, %%rax\n  "
+    "mulq %10\n  "
+    "addq %%rax, %2\n  "
+    "adcq $0, %%rdx\n  "
+    "movq %%rdx, %3\n  "
+
+    "movq %4, %%rax\n  "
+    "mulq %11\n  "
+    "addq %%rax, %3\n  "
+    "adcq $0, %%rdx\n  "
+    "movq %%rdx, %%rcx\n  "
+
+    "movq %5, %%rax\n  "
+    "mulq %8\n  "
+    "addq %%rax, %1\n  "
+    "adcq %%rdx, %2\n  "
+    "adcq $0, %3\n  "
+    "adcq $0, %%rcx\n  "
+
+    "movq %5, %%rax\n  "
+    "mulq %9\n  "
+    "addq %%rax, %2\n  "
+    "adcq %%rdx, %3\n  "
+    "adcq $0, %%rcx\n  "
+    "adcq $0, %3\n  "
+    "adcq $0, %%rcx\n  "
+
+    "movq %5, %%rax\n  "
+    "mulq %10\n  "
+    "addq %%rax, %3\n  "
+    "adcq $0, %%rdx\n  "
+    "orq %%rdx, %%rcx\n  "
+
+    "movq %6, %%rax\n  "
+    "mulq %8\n  "
+    "addq %%rax, %2\n  "
+    "adcq %%rdx, %3\n  "
+    "adcq $0, %%rcx\n  "
+
+    "movq %6, %%rax\n  "
+    "mulq %9\n  "
+    "addq %%rax, %3\n  "
+    "adcq $0, %%rdx\n  "
+    "orq %%rdx, %%rcx\n  "
+
+    "movq %7, %%rax\n  "
+    "mulq %8\n  "
+    "addq %%rax, %3\n  "
+    "orq %%rdx, %%rcx\n  "
+
+    "cmpq $0, %%rcx\n  "
+    "jne 2f\n  "
+
+    "movq %7, %%rcx\n  "
+    "jrcxz 12f\n  "
+
+    "movq %11, %%rcx\n  "
+    "movq %10, %%rax\n  "
+    "orq %%rax, %%rcx\n  "
+    "movq %9, %%rax\n  "
+    "orq %%rax, %%rcx\n  "
+    "jmp 2f\n  "
+
+    "12:\n  "
+    "movq %11, %%rcx\n  "
+    "jrcxz 11f\n  "
+
+    "movq %6, %%rcx\n  "
+    "movq %5, %%rax\n  "
+    "or %%rax, %%rcx\n  "
```

Comment:
why is this `or` where in other places are `orq`, do you need to use `orq` explicitly at all?

Reply:
No, `orq` isn't necessary. I added the type suffixes in because originally it looked like it was failing on values that are a multiple of `2<<32` and I thought there was a cast happening (turns out it was unrelated, GCC was allocating overlapping registers, it was just that the number generation code was generating numbers that weren't very random). I'll just use the unsuffixed versions everywhere.